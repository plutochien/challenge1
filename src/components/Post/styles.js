import styled from 'styled-components';

export const PostContainer = styled.div`
  padding: 10px 15px;
  display: flex;
  flex-direction: column;
  border-bottom: 2px solid rgb(230, 236, 240);
`;

export const PostTitle = styled.div`
  font-weight: bold;
  font-size: 15px;
`;
export const PostBody = styled.div`
  font-size: 15px;
  padding: 10px;
`;

export const Button = styled.button`
  margin-left: auto;
  margin-top: 5px;
  padding: 5px 5px 0 5px;
  border: none;
  background: none;
  outline: none;
  &:hover {
    background: rgb(224, 239, 250);
    border-radius: 25px;
    -webkit-transition: background-color 0s;
    transition: background-color 0s;
  }
`;
