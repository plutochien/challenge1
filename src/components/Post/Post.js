import React, { Fragment, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { PostContainer, PostTitle, PostBody, Button } from './styles';
import ArrowIcon from '../../assets/icons/expand_arrow.svg';
import { bindActionCreators } from 'redux';
import { fetchComments } from '../../actions/commentAction';
import { connect } from 'react-redux';
import Spinner from '../Spinner/Spinner';
import Comment from '../Comment/Comment';

const Post = ({
  post: { id, title, body },
  comments,
  fetchComments,
  currentPostId,
  isCommentLoading,
}) => {
  const [showComment, setShowCommet] = useState(false);
  useEffect(() => {
    if (currentPostId !== id) {
      setShowCommet(false);
    }
  }, [currentPostId, id]);

  const handleClick = event => {
    event.preventDefault();

    fetchComments(id);
    setShowCommet(!showComment);
  };

  const renderShowComment = () => {

    if (!showComment || isCommentLoading) {
      return (
        <Button onClick={handleClick}>
          <img src={ArrowIcon} alt="expand" />
        </Button>
      );
    }
    return (
      <Fragment>
        {comments.map(comment => (
          <Comment key={comment.id} comment={comment} postId={id} />
        ))}
        <Button onClick={handleClick}>
          <img
            src={ArrowIcon}
            style={{ transform: 'scaleY(-1)' }}
            alt="collapse"
          />
        </Button>
      </Fragment>
    );
  };
  return (
    <PostContainer>
      {isCommentLoading && <Spinner />}
      <PostTitle>{title}</PostTitle>
      <PostBody>{body}</PostBody>
      {renderShowComment()}
    </PostContainer>
  );
};

Post.propTypes = {
  post: PropTypes.shape({
    id: PropTypes.number.isRequired,
    body: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
  }).isRequired,
  comments: PropTypes.array.isRequired,
  fetchComments: PropTypes.func.isRequired,
  currentPostId: PropTypes.number,
  isCommentLoading: PropTypes.bool,
};

const mapStateToProps = (state, ownProps) => ({
  comments: state.commentsReducer.comments,
  isCommentLoading: state.usersReducer.isCommentLoading,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      fetchComments,
    },
    dispatch,
  );

export default connect(mapStateToProps, mapDispatchToProps)(Post);
