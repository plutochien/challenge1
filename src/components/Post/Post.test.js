import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Post from './Post';
const mockStore = configureStore([]);

describe('Post Component', () => {
  let store;
  let component;
  beforeEach(() => {
    store = mockStore({
      usersReducer: {
        users: [
          {
            id: 1,
            name: 'Leanne Graham',
            username: 'Bret',
            email: 'Sincere@april.biz',
            phone: '1-770-736-8031 x56442',
            website: 'hildegard.org',
            company: {
              name: 'Romaguera-Crona',
              catchPhrase: 'Multi-layered client-server neural-net',
              bs: 'harness real-time e-markets',
            },
          },
        ],
        isUserLoading: false,
      },
      postsReducer: {
        posts: [
          {
            userId: 4,
            id: 31,
            title: 'ullam ut quidem id aut vel consequuntur',
            body: 'debitis eius sed qadipisci repudiandae',
          },
        ],
        isPostLoading: false,
      },
      commentsReducer: {
        comments: [
          {
            postId: 32,
            id: 156,
            name: 'pariatur omnis in',
            email: 'Telly_Lynch@karl.co.uk',
            body: 'dolorum voluptas laborinus dolores magni',
          },
        ],
        isCommentsLoading: false,
        currentPostId: 31,
      },
    });
    store.dispatch = jest.fn();
    component = renderer.create(
      <Provider store={store}>
        <Post post={{ id: 31, title: 'test title', body: 'test body' }} />
      </Provider>,
    );
  });

  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
