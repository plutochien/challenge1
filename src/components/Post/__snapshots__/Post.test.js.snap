// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Component should render with given state from Redux store 1`] = `
<div
  className="sc-bdfBwQ exZruV"
>
  <div
    className="sc-gsTCUz iPsnom"
  >
    test title
  </div>
  <div
    className="sc-dlfnbm bfzwnW"
  >
    test body
  </div>
  <button
    className="sc-hKgILt khksFS"
    onClick={[Function]}
  >
    <img
      alt="expand"
      src="expand_arrow.svg"
    />
  </button>
</div>
`;
